pkg/providers/openstack/delete_image.go

```diff
@@ -2,8 +2,55 @@ package openstack
 
 import (
 	"github.com/emc-advanced-dev/pkg/errors"
+	"github.com/emc-advanced-dev/unik/pkg/types"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack/imageservice/v2/images"
 )
 
 func (p *OpenstackProvider) DeleteImage(id string, force bool) error {
-	return errors.New("not yet supportded for openstack", nil)
+	image, err := p.GetImage(id)
+	if err != nil {
+		return errors.New("retrieving image", err)
+	}
+
+	// Delete instances of this image.
+	instances, err := p.ListInstances()
+	if err != nil {
+		return errors.New("retrieving list of instances", err)
+	}
+	for _, instance := range instances {
+		if instance.ImageId == image.Id {
+			if !force {
+				return errors.New("instance "+instance.Id+" found which uses image "+image.Id+"; try again with force=true", nil)
+			} else {
+				err = p.DeleteInstance(instance.Id, true)
+				if err != nil {
+					return errors.New("failed to delete instance "+instance.Id+" which is using image "+image.Id, err)
+				}
+			}
+		}
+	}
+
+	clientGlance, err := p.newClientGlance()
+	if err != nil {
+		return err
+	}
+
+	if err := deleteImage(clientGlance, image.Id); err != nil {
+		return errors.New("failed deleting image "+image.Id, err)
+	}
+
+	// Update state.
+	if err := p.state.ModifyImages(func(imageList map[string]*types.Image) error {
+		delete(imageList, image.Id)
+		return nil
```

Comment:
Your state is actually state.BasicState object. In the daemon code, where you initialise the provider it will us empty state if it doesn't exist, e.g.

https://github.com/emc-advanced-dev/unik/pull/78/commits/b10a9b05a34edbde4c7bf49c565fcb65eb49e3bb#diff-06b532a2a5a5d3ad1d5341b3e00945efR163

Saving of the state after modifying images (or other objects) is therefore provided by the BasicState at

https://github.com/emc-advanced-dev/unik/blob/master/pkg/state/base_state.go#L86